File: qbopt/frontend/qb/stage_text.py
"""Diagnostic MASM/Intel projection for virtual and allocated LIR.

This is deliberately outside the backend: it prints pass state and does not
participate in selection, allocation, encoding, or object emission.
"""

from iced_x86 import Register
from iced_x86 import Register_

from qbopt.model import ir
from qbopt.backend import target
from qbopt.objectfile.module import Addr
from qbopt.objectfile.module import Space

SIZES = {1: "byte", 2: "word", 4: "dword", 8: "qword", 10: "tbyte"}


def instruction_text(what: ir.Semantics) -> tuple[str, ...]:
    """Render one LIR operation in Intel operand order and MASM spelling."""
    name = what.name or ""
    dests = tuple(_operand(one) for one in what.dests)
    sources = tuple(_operand(one) for one in what.sources)
    match what.op:
        case ir.Operation.NOTHING:
            return (name,) if name not in ("", "nop") else ()
        case ir.Operation.MOVE | ir.Operation.ADDRESS:
            return (f"{name} {dests[0]}, {sources[0]}",)
        case ir.Operation.BINARY:
            return (f"{name} {dests[0]}, {sources[1]}",)
        case ir.Operation.UNARY:
            return (f"{name} {dests[0]}",)
        case ir.Operation.COMPARE if name.startswith("f"):
            memory = [text for text, source in zip(sources, what.sources, strict=True) if isinstance(source, ir.Mem)]
            return (f"{name} {memory[0]}" if memory else name,)
        case ir.Operation.COMPARE:
            return (f"{name or 'cmp'} {sources[0]}, {sources[1]}",)
        case ir.Operation.MULTIPLY if len(dests) == 1:
            if len(sources) == 3:
                return (f"imul {dests[0]}, {sources[1]}, {sources[2]}",)
            if isinstance(what.sources[1], ir.Imm):
                return (f"imul {dests[0]}, {sources[0]}, {sources[1]}",)
            return (f"imul {dests[0]}, {sources[1]}",)
        case ir.Operation.MULTIPLY | ir.Operation.DIVIDE:
            return (f"{name} {sources[-1]}",)
        case ir.Operation.EXTEND:
            return (f"{name} {dests[0]}, {sources[0]}" if name in ("movsx", "movzx") else name,)
        case ir.Operation.PUSH:
            suffix = ""
            if isinstance(what.sources[0], ir.Imm):
                suffix = "d" if what.sources[0].width == 4 else "w"
            return (f"push{suffix} {sources[0]}",)
        case ir.Operation.POP:
            return (f"pop {dests[0]}",)
        case ir.Operation.EXCHANGE if name == "fxch":
            return (f"fxch {dests[1]}",)
        case ir.Operation.EXCHANGE:
            return (f"xchg {dests[0]}, {dests[1]}",)
        case ir.Operation.FUNNEL:
            return (f"{name} {dests[0]}, {sources[1]}, {sources[2]}",)
        case ir.Operation.BRANCH | ir.Operation.JUMP:
            return (f"{name} L0_{what.target}",)
        case ir.Operation.CALL if what.indirect and sources:
            return (f"call {sources[0]}",)
        case ir.Operation.CALL:
            return (f"call {name}",)
        case ir.Operation.FILL:
            return (f"rep {name}",)
        case ir.Operation.BARRIER:
            operands = dests or sources
            return (f"{name} {operands[0]}" if operands else name,)
        case ir.Operation.FLOAT_LOAD:
            return (name if name in ("fldz", "fld1") or not sources else f"{name} {sources[0]}",)
        case ir.Operation.FLOAT_STORE:
            operands = dests or sources
            return (f"{name} {operands[0]}" if operands else name,)
        case ir.Operation.FLOAT_ARITH if isinstance(what.sources[-1], ir.Mem):
            return (f"{name} {sources[-1]}",)
        case ir.Operation.FLOAT_ARITH | ir.Operation.FLOAT_ARITH_POP:
            return (f"{name} {dests[0]}, {sources[-1]}",)
        case ir.Operation.FLOAT_UNARY | ir.Operation.LEAVE | ir.Operation.RETURN:
            return (name,) if name else ()
        case ir.Operation.ESCAPE:
            return (f"jmp far ptr {sources[0] if sources else name}",)
        case ir.Operation.RESTORE:
            return (f"restore {', '.join((*dests, *sources))}",)
        case ir.Operation.DATA:
            return (name or "db ?",)
    return (f"; unprintable {what.op.value} {name}",)


def inline_text(parts: tuple[object, ...]) -> tuple[str, ...]:
    """Render a finalized in-place intrinsic body as MASM data directives."""
    lines: list[str] = []
    for part in parts:
        match part:
            case bytes():
                lines.extend(
                    "db " + ",".join(f"0{byte:02x}h" for byte in part[start : start + 16])
                    for start in range(0, len(part), 16)
                )
            case ("offset", str(name), int(offset)):
                lines.append(f"dw offset {name}{_signed(offset)}")
            case ("segment", str(name), _):
                lines.append(f"dw seg {name}")
    return tuple(lines)


def _operand(where: ir.Loc) -> str:
    match where:
        case ir.Reg(register=register):
            return target.name_of(register)
        case ir.Held(value=value):
            return f"v{value}"
        case ir.St(index=index):
            return f"st({index})"
        case ir.Imm(value=value, address=None):
            return str(value)
        case ir.Imm(value=value, address=address):
            if address.space is Space.GROUP:
                return _symbol(address)
            return f"offset {_symbol(address)}{_signed(address.disp + value)}"
        case ir.Mem():
            return _memory(where)
        case ir.Address(addr=address, index=Register.NONE) if address is not None:
            return _memory(ir.Mem(address, 2)).removeprefix("word ptr ")
        case ir.Address(through=through, index=index, scale=scale, offset=offset):
            return f"[{_registers(through, index, scale)}{_signed(offset)}]"
    return f"<{type(where).__name__.lower()}>"


def _memory(cell: ir.Mem) -> str:
    size = f"{SIZES.get(cell.width, f'{cell.width}-byte')} ptr "
    address = cell.addr
    if cell.base is not None or cell.index is not None or cell.selector is not None:
        registers = []
        if cell.base is not None:
            registers.append(f"v{cell.base.value}")
        if cell.index is not None:
            registers.append(f"v{cell.index.value}" + (f"*{cell.scale}" if cell.scale != 1 else ""))
        inside = "+".join(registers)
        displacement = cell.offset if address is None else address.disp
        if address is not None and address.space in (Space.SEGMENT, Space.EXTERNAL, Space.GROUP):
            inside = _symbol(address) + (("+" + inside) if inside else "")
        selector = f"v{cell.selector.value}:" if cell.selector is not None else ""
        return f"{size}{selector}[{inside}{_signed(displacement)}]"
    if address is None:
        base = target.name_of(cell.through) if cell.through != Register.NONE else "?"
        return f"{size}[{base}{_signed(cell.offset)}]"
    registers = _registers(cell.through, cell.index_through, cell.scale)
    disp = _signed(address.disp)
    match address.space:
        case Space.FRAME:
            return f"{size}[bp{disp}]"
        case Space.SEGMENT | Space.EXTERNAL | Space.GROUP:
            return f"{size}{_symbol(address)}{disp}" + (f"[{registers}]" if registers else "")
        case Space.LITERAL:
            segment = "" if address.segment == Register.NONE else f"{target.name_of(address.segment)}:"
            return f"{size}{segment}[{registers}{disp}]"
        case Space.FAR:
            return f"{size}{target.name_of(address.segment)}:[{registers}{disp}]"
        case Space.STACK:
            return f"{size}[sp{disp}]"
    return f"{size}[?]"


def _registers(base: Register_, index: Register_, scale: int = 1) -> str:
    parts = [target.name_of(base)] if base != Register.NONE else []
    if index != Register.NONE:
        parts.append(target.name_of(index) + (f"*{scale}" if scale != 1 else ""))
    return "+".join(parts)


def _symbol(address: Addr) -> str:
    return f"{address.space.value}_{address.index}"


def _signed(number: int) -> str:
    return f"+{number}" if number > 0 else (str(number) if number < 0 else "")

File: src/frontends/qb/stage_text.rs
//! Port of `qbopt/frontend/qb/stage_text.py`: diagnostic MASM/Intel
//! projection for virtual and allocated LIR.
//!
//! Outside the backend: it prints pass state and does not participate in
//! selection, allocation, encoding, or object emission.

use iced_x86::Register;

use crate::backend::masm::InlinePart;
use crate::backend::target;
use crate::model::ir::{self, Loc, Operation};
use crate::objectfile::module::{Addr, Space};

fn size_name(width: u32) -> String {
    match width {
        1 => "byte".into(),
        2 => "word".into(),
        4 => "dword".into(),
        8 => "qword".into(),
        10 => "tbyte".into(),
        _ => format!("{width}-byte"),
    }
}

/// Render one LIR operation in Intel operand order and MASM spelling.
pub fn instruction_text(what: &ir::Semantics) -> Vec<String> {
    let name = what.name.clone().unwrap_or_default();
    let dests: Vec<String> = what.dests.iter().map(_operand).collect();
    let sources: Vec<String> = what.sources.iter().map(_operand).collect();
    let one = |text: String| vec![text];
    match what.op {
        Operation::Nothing => {
            if name != "" && name != "nop" {
                one(name)
            } else {
                vec![]
            }
        }
        Operation::Move | Operation::Address => one(format!("{name} {}, {}", dests[0], sources[0])),
        Operation::Binary => one(format!("{name} {}, {}", dests[0], sources[1])),
        Operation::Unary => one(format!("{name} {}", dests[0])),
        Operation::Compare if name.starts_with('f') => {
            let memory: Vec<&String> = sources
                .iter()
                .zip(&what.sources)
                .filter(|(_, source)| matches!(source, Loc::Mem(_)))
                .map(|(text, _)| text)
                .collect();
            one(if let Some(first) = memory.first() { format!("{name} {first}") } else { name })
        }
        Operation::Compare => {
            one(format!("{} {}, {}", if name.is_empty() { "cmp" } else { name.as_str() }, sources[0], sources[1]))
        }
        Operation::Multiply if dests.len() == 1 => {
            if sources.len() == 3 {
                return one(format!("imul {}, {}, {}", dests[0], sources[1], sources[2]));
            }
            if matches!(what.sources[1], Loc::Imm(_)) {
                return one(format!("imul {}, {}, {}", dests[0], sources[0], sources[1]));
            }
            one(format!("imul {}, {}", dests[0], sources[1]))
        }
        Operation::Multiply | Operation::Divide => one(format!("{name} {}", sources[sources.len() - 1])),
        Operation::Extend => one(if name == "movsx" || name == "movzx" {
            format!("{name} {}, {}", dests[0], sources[0])
        } else {
            name
        }),
        Operation::Push => {
            let mut suffix = "";
            if let Loc::Imm(imm) = &what.sources[0] {
                suffix = if imm.width == 4 { "d" } else { "w" };
            }
            one(format!("push{suffix} {}", sources[0]))
        }
        Operation::Pop => one(format!("pop {}", dests[0])),
        Operation::Exchange if name == "fxch" => one(format!("fxch {}", dests[1])),
        Operation::Exchange => one(format!("xchg {}, {}", dests[0], dests[1])),
        Operation::Funnel => one(format!("{name} {}, {}, {}", dests[0], sources[1], sources[2])),
        Operation::Branch | Operation::Jump => {
            one(format!("{name} L0_{}", what.target.map_or_else(|| "None".to_owned(), |one| one.to_string())))
        }
        Operation::Call if what.indirect && !sources.is_empty() => one(format!("call {}", sources[0])),
        Operation::Call => one(format!("call {name}")),
        Operation::Fill => one(format!("rep {name}")),
        Operation::Barrier => {
            let operands = if dests.is_empty() { &sources } else { &dests };
            one(if let Some(first) = operands.first() { format!("{name} {first}") } else { name })
        }
        Operation::FloatLoad => one(if name == "fldz" || name == "fld1" || sources.is_empty() {
            name
        } else {
            format!("{name} {}", sources[0])
        }),
        Operation::FloatStore => {
            let operands = if dests.is_empty() { &sources } else { &dests };
            one(if let Some(first) = operands.first() { format!("{name} {first}") } else { name })
        }
        Operation::FloatArith if matches!(what.sources.last(), Some(Loc::Mem(_))) => {
            one(format!("{name} {}", sources[sources.len() - 1]))
        }
        Operation::FloatArith | Operation::FloatArithPop => {
            one(format!("{name} {}, {}", dests[0], sources[sources.len() - 1]))
        }
        Operation::FloatUnary | Operation::Leave | Operation::Return => {
            if name.is_empty() {
                vec![]
            } else {
                one(name)
            }
        }
        Operation::Escape => one(format!(
            "jmp far ptr {}",
            if let Some(first) = sources.first() { first.clone() } else { name }
        )),
        Operation::Restore => {
            let all: Vec<String> = dests.iter().chain(&sources).cloned().collect();
            one(format!("restore {}", all.join(", ")))
        }
        Operation::Data => one(if name.is_empty() { "db ?".to_owned() } else { name }),
        // Python's `; unprintable {op} {name}` fallback: every operation is matched above.
    }
}

/// Render a finalized in-place intrinsic body as MASM data directives.
pub fn inline_text(parts: &[InlinePart]) -> Vec<String> {
    let mut lines = Vec::new();
    for part in parts {
        match part {
            InlinePart::Bytes(bytes) => {
                for chunk in bytes.chunks(16) {
                    lines.push(
                        "db ".to_owned()
                            + &chunk.iter().map(|byte| format!("0{byte:02x}h")).collect::<Vec<_>>().join(","),
                    );
                }
            }
            InlinePart::Fixup(kind, name, offset) if kind == "offset" => {
                lines.push(format!("dw offset {name}{}", _signed(*offset)));
            }
            InlinePart::Fixup(kind, name, _) if kind == "segment" => lines.push(format!("dw seg {name}")),
            InlinePart::Fixup(..) => {}
        }
    }
    lines
}

fn _operand(r#where: &Loc) -> String {
    match r#where {
        Loc::Reg(reg) => target::name_of(reg.register),
        Loc::Held(held) => format!("v{}", held.value),
        Loc::St(st) => format!("st({})", st.index),
        Loc::Imm(ir::Imm { value, address: None, .. }) => value.to_string(),
        Loc::Imm(ir::Imm { value, address: Some(address), .. }) => {
            if address.space == Space::Group {
                return _symbol(address);
            }
            format!("offset {}{}", _symbol(address), _signed(address.disp + value))
        }
        Loc::Mem(cell) => _memory(cell),
        Loc::Address(address) if address.index == Register::None && address.addr.is_some() => {
            let text = _memory(&ir::Mem::new(address.addr, 2));
            text.strip_prefix("word ptr ").map_or(text.clone(), str::to_owned)
        }
        Loc::Address(address) => {
            format!("[{}{}]", _registers(address.through, address.index, address.scale), _signed(address.offset))
        }
    }
}

fn _memory(cell: &ir::Mem) -> String {
    let size = format!("{} ptr ", size_name(cell.width));
    let address = cell.addr;
    if cell.base.is_some() || cell.index.is_some() || cell.selector.is_some() {
        let mut registers = Vec::new();
        if let Some(base) = cell.base {
            registers.push(format!("v{}", base.value));
        }
        if let Some(index) = cell.index {
            registers.push(format!("v{}", index.value) + &if cell.scale != 1 { format!("*{}", cell.scale) } else { String::new() });
        }
        let mut inside = registers.join("+");
        let displacement = address.map_or(cell.offset, |address| address.disp);
        if let Some(address) = address {
            if matches!(address.space, Space::Segment | Space::External | Space::Group) {
                inside = _symbol(&address) + &if inside.is_empty() { String::new() } else { format!("+{inside}") };
            }
        }
        let selector = cell.selector.map_or_else(String::new, |selector| format!("v{}:", selector.value));
        return format!("{size}{selector}[{inside}{}]", _signed(displacement));
    }
    let Some(address) = address else {
        let base = if cell.through != Register::None { target::name_of(cell.through) } else { "?".to_owned() };
        return format!("{size}[{base}{}]", _signed(cell.offset));
    };
    let registers = _registers(cell.through, cell.index_through, cell.scale);
    let disp = _signed(address.disp);
    match address.space {
        Space::Frame => format!("{size}[bp{disp}]"),
        Space::Segment | Space::External | Space::Group => format!(
            "{size}{}{disp}{}",
            _symbol(&address),
            if registers.is_empty() { String::new() } else { format!("[{registers}]") }
        ),
        Space::Literal => {
            let segment = if address.segment == Register::None {
                String::new()
            } else {
                format!("{}:", target::name_of(address.segment))
            };
            format!("{size}{segment}[{registers}{disp}]")
        }
        Space::Far => format!("{size}{}:[{registers}{disp}]", target::name_of(address.segment)),
        Space::Stack => format!("{size}[sp{disp}]"),
    }
}

fn _registers(base: Register, index: Register, scale: i64) -> String {
    let mut parts = if base != Register::None { vec![target::name_of(base)] } else { vec![] };
    if index != Register::None {
        parts.push(target::name_of(index) + &if scale != 1 { format!("*{scale}") } else { String::new() });
    }
    parts.join("+")
}

fn _symbol(address: &Addr) -> String {
    format!("{}_{}", address.space.value(), address.index)
}

fn _signed(number: i64) -> String {
    if number > 0 {
        format!("+{number}")
    } else if number < 0 {
        number.to_string()
    } else {
        String::new()
    }
}
